### ISSUE TITLE
Add auto-generated test files

### ISSUE THREAD

**Issue by wing328** [swagger-api/swagger-codegen#2276]
As part of https://github.com/swagger-api/swagger-codegen/pull/1921, we've add capabilities to generate test files for API and model files. 

If anyone has cycle to add the feature to the following languages, please reply to let us know:

- [ ] Android
- [ ] C++
- [x] C# (#2007)
- [ ] Clojure
- [ ] Dart
- [ ] Flash
- [ ] Javascript
- [ ] Java
- [ ] ObjC
- [x] Perl (#2109)
- [x] PHP (#1921)
- [ ] Python (#2161)
- [ ] Ruby (#2272)
- [ ] TypeScript (#2270)
- [ ] Swift
- [ ] Scala

**Comment by ePaul** [comment]
I don't quite understand: What is the purpose of those generated tests?
The PHP tests seem to be empty, so they are not yet testing anything – It looks as after generation they need to be edited to actually test something. But then the "don't edit this file" comment should go away.

But then what is being tested here? The getters/setters of the model objects? (That kind of test, if deemed necessary, should also be generated. Though I would (as a codegen-user, not as a maintainer of codegen) prefer if the generated code just works and I don't have to test it myself.)

What might be useful would be a test of the actual API methods (which is to be run from a test framework against an actual server) – generating some template as a start for your own written tests.

**Comment by wing328** [comment]
The purpose is to save developer's time in creating those test files from scratch. (I'll remove "don't edit this file" in my next PR)

And yes, ideally we should test getter/setter for each model. The work I've done is just a starting point. Ideally the test fils should leverage all available information provided in the spec to generate all possible test cases.

In my opinion, generating the test code makes the SDK/API client looks more mature. For example, if there are 2 open-source Ruby SDKs for Twitter API - one with test cases and another without any, I would tend to think the one with test cases is more mature. What do you think? 

There are scaffolded test cases for API methods, e.g. https://github.com/swagger-api/swagger-codegen/blob/master/samples/client/petstore/csharp/SwaggerClientTest/Lib/SwaggerClient.Test/PetApiTests.cs. Is that what you're looking for?

Thanks again for the feedback. I think what I'll do it o update the wording to elaborate more on the purpose of the scaffolded test cases in the auto-generated test files.

**Comment by wing328** [comment]
@ePaul does it make sense to you?

Later if anyone from the community has cycle, we can update codegen to automatically generate test cases specified by vendor extension in the endpoint definition.

**Comment by wing328** [comment]
@ePaul FYI. https://github.com/swagger-api/swagger-codegen/pull/2301 to revise the wording has been merged into master.

**Comment by ePaul** [comment]
@wing328 Thanks, this way it is clearer.

**Comment by waiterZen** [comment]
Hi, I want to contribute javascript related test code. And I also want to add some feature after driven down the core if I find something can make improvement. If I choose JS, all the pr will commit to this [reps](https://github.com/wcandillon/swagger-js-codegen)?

**Comment by wing328** [comment]
@waiterZen that's a separate project. You can refer to the PRs above to get a better understanding of what the change (adding auto-generated test files) looks like

**Comment by mhuisman** [comment]
in java/retrofit2 there is sufficient functionality in `ApiClient.mustache` to merit adding a default template that offers unit tests for that functionality.  I'm shoehorning it into into `api_test.mustache` but it feels imprecise/ugly.  Would you be interested in a PR addressing adding `ApiClient_test.mustache`?

**Comment by wing328** [comment]
>  Would you be interested in a PR addressing adding ApiClient_test.mustache?

Definitely. Please kindly file a PR so that we can review the enhancement. Thanks.

**Comment by wing328** [comment]
@mhuisman if you need help with the PR, please reply to let us know.

**Comment by cbornet** [comment]
In that case please do it against branch 2.3.0 since there were some changes in the API.

**Comment by grokify** [comment]
> [ ] Go (https://github.com/swagger-api/swagger-codegen/pull/2532, https://github.com/swagger-api/swagger-codegen/pull/2577)

What else is necessary for Go to get a ☑?
